Add 'Writing Apache Beam Jobs' section to wiki (#43)
U8NWXD/oppia-wiki: Apache-Beam-Jobs.md

## Table of contents

-   [Introduction](#introduction)
-   [Apache Beam Job Architecture](#apache-beam-job-architecture)
    -   [`Pipeline`s](#pipelines)
    -   [`PValue`s](#pvalues)
    -   [`PTransform`s](#ptransforms)
        -   [`ParDo` and `DoFn`](#pardo-and-dofn)
        -   [`Map` and `FlatMap`](#map-and-flatmap)
        -   [`Filter`](#filter)
        -   [`GroupByKey`](#groupbykey)
            -   [Example of using `GroupByKey`,`Filter`, and `FlatMap`](#example-of-using-groupbykeyfilter-and-flatmap)
    -   [`Runner`s](#runners)
-   [Writing Apache Beam Jobs](#writing-apache-beam-jobs)
-   [Testing Apache Beam Jobs](#testing-apache-beam-jobs)
-   [Running Apache Beam Jobs](#running-apache-beam-jobs)
-   [Case studies](#case-studies)

>     -   [Case Study: `SchemaMigrationJob`](#case-study-schemamigrationjob)

Reviewer: Can you preserve the original format (e.g. with asterisks, not so many spaces after each bullet)? The original table of contents was auto-generated

Author: Done.

`Runner`s provide the `run()` method used to visit every node (`PValue`) in the pipeline's DAG by executing the edges (`PTransform`s) to compute their values.  At Oppia, we use `DataflowRunner` to have our `Pipeline`s run on the [Google Cloud Dataflow service](https://cloud.google.com/dataflow).

## Writing Apache Beam Jobs

For this section, we'll walk through the steps of implementing a job by writing one: `CountExplorationStatesJob`.

It's helpful to begin by sketching a diagram of what you want the job to do. We recommend using pen and paper or a whiteboard, but in this wiki page we'll use ASCII art to keep the document self-contained.

Here's a diagram for the `CountExplorationStatesJob`:

    .--------------. Count states .--------. Sum .-------.
    | Explorations | -----------> | Counts | --> | Total |
    '--------------'              '--------'     '-------'

> **TIP**: As illustrated, you don't need to know what the names of the `PTransform`s (edges) used in a diagram are. It's easy to look up the appropriate `PTransform` after drawing the diagram.

Now that we have our bearings, let's get started on implementing the job.

> ### 1. Subclass the `base_jobs.JobBase` class and override the `run()` method

Reviewer: Can you update the table of contents to include this step? We've included other level-3 headings in the TOC

Author: Done.

Make sure your job class name is clear and concise, because the name is presented to release coordinators:

> ![image](https://user-images.githubusercontent.com/5094060/135734501-eb9c0370-e98d-4271-b41a-0bf11c25503c.png)

Reviewer: I was thinking about something more descriptive than "image." Could the alt text more clearly explain what the image shows?

Author: Whoops, this was more descriptive but it got deleted. Fixed!

        'e2', 'o1', ['D', 'E', 'F', 'G', 'H'], ['TextInput'])
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e3', 'o1', ['I', 'J'], ['TextInput'])
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e4', 'o1', ['K', 'L', 'M', 'N'], ['TextInput'])

    self.assert_job_output_is([
        job_run_result.JobRunResult(stdout='14'),
    ])
```

Note that `self.assert_job_output_is(...)` and `self.assert_job_output_is_empty()` do as advertised -- they run the job to completion and verify the result.

> **IMPORTANT:** Only one `assert_job_output_is` assertion can be performed in a test body. Multiple calls will result in an exception instructing you to split the test apart.

Just because a job passes in unit tests does not guarantee it will pass in production. This is because workers, which execute the pipeline code, are run in a special environment where the code base is configured differently. While Oppia's jobs team works to resolve the differences, be careful about using complex and/or confusing objects. The simpler your job, the greater chance it'll work in production!

## Running Apache Beam Jobs

### Local Development Server

> These instructions assume you are running a local development server. If you are a release coordinator running these jobs on the production or testing servers, you should already have been granted the "Release Coordinator" role, so you can skip steps 1-3.

Reviewer 2: Should we mention somewhere that Beam validates Python types and that they shouldn't be left as the last step?

Author: Done.

1. Sign in as an administrator ([instructions][3]).
2. Navigate to **Admin Page > Roles Tab**.
3. Add the "Release Coordinator" role to the username you are signed in with.
4. Navigate to http://localhost:8181/release-coordinator, then to the **Beam Jobs tab**.
5. Search for your job and then click the **Play button**.
6. Click "Start new job".

![Screen recording showing how to run jobs](https://user-images.githubusercontent.com/5094060/128743997-70cca5f9-0b76-4294-806e-f65f5df5be95.gif)

### Production Server

Before a job or feature that cannot be fully tested locally can be run and deployed in production, it must first be tested on the Oppia backup server. Examples of features that require backup server testing are features that add or modify code that requires some third-party API, like Cloud Tasks, Cloud Storage, or Cloud Translate.

If your job or feature is not essential for the release and has not been fully tested by the release cut, then it is not going into the release. "Fully tested" means:
- The job or feature should run without failures on the Oppia backup server.
- The job or feature produces the expected output.

> - The job or feature has the expected outcome (this must be verified by e.g. user-facing changes, or introspection of the datastore, or a validation job, or an output check, etc.).

Reviewer 2: We should not use datastore introspection except for rare cases where it is the only possible solution.

Author: OK, I've omitted it.

- The job or feature should be explicitly approved by server jobs admin (currently @seanlip and @vojtechjelinek).

> Also, in case your job is a new migration job (not any already existing migrations), there has to be an audit job accompanying it to verify that the data that you are migrating is valid in the server. **The audit job will have to be "Fully tested" before testing on the migration job can start.**

Reviewer 2: ```suggestion
Also, in case your job changes data in the datastore, there has to be a validation job accompanying it to verify that the data that you are migrating is valid in the server. **The validation job will have to be "Fully tested" before testing on the migration job can start.**
```

Author: Done.

`Runner`s provide the `run()` method used to visit every node (`PValue`) in the pipeline's DAG by executing the edges (`PTransform`s) to compute their values.  At Oppia, we use `DataflowRunner` to have our `Pipeline`s run on the [Google Cloud Dataflow service](https://cloud.google.com/dataflow).

> ## Writing Apache Beam Jobs

Reviewer: I feel like this section needs more setup so that a contributor who is new to beam jobs doesn't get lost

Author: Done.

The name of your job class is presented to release coordinators, so make sure it is clear and concise:
![Screenshot from 2021-09-28 09-05-10](https://user-images.githubusercontent.com/5094060/135092574-0bff536e-1b58-4b38-9358-c26f9096c8a3.png)

**Job names should follow the convention: `<Verb><Noun>Job`.**

For example:

```python
class WeeklyDashboardStatsComputationJob(base_jobs.JobBase):
    """BAD: Name does not begin with a verb."""

    def run(self):
        ...

class ComputeStatsJob(base_jobs.JobBase):
    """BAD: Unclear what kind of stats are being computed."""

    def run(self):
        ...

class CountExplorationInteractionsJob(base_jobs.JobBase):
    """GOOD: Name starts with a verb and "exploration interactions" is unambiguous."""

    def run(self):
        ...
```

**Module names should follow the convention: `<entity>_<operation>_jobs.py`.**

For example:
* `blog_validation_jobs.py`
* `dashboard_stats_computation_jobs.py`
* `exploration_indexing_jobs.py`
* `exploration_stats_regeneration_jobs.py`
* `model_validation_jobs.py`

**New modules must be imported in the `jobs/registry.py` file.**

---

> The `run()` method must return a `PCollection[JobRunResult]`.

Reviewer: I think it would be clearer to say "PCollection of JobRunResults" since not everyone is going to be familiar with the `PCollection[JobRunResult]` syntax

* In English, this means that **the job _must_ report _something_ about what occurred during its execution.** For example, this can be the errors it discovered or the number of successful operations it was able to perform. **Empty results are forbidden!**

  * If you don't think your job has any results worth reporting, then just print a "success" metric with the number of models it processed.

* `JobRunResult` outputs should answer the following questions:

  * Did the job run without any problems? How and why do I know?
  * How much work did the job manage to do?

>   * If the job encountered a problem, what caused it?

Reviewer: Should we remove the duplicate of this content in the case study below?

Author: Done.

> When implementing the `run()` method, make liberal usage of small and simple `PTransform`s and `DoFn`s. You must also be careful to only use the following legal constructs:

Reviewer: > You must also be careful to only use the following legal constructs:

This makes it sound like the items in the table below are the only allowed PTransforms in jobs. Do you mean that these are the only allowed ways to get, put, and delete models?

Author: How's this?

| Operation       | Do use                                 | Do not use                     |
| --------------- | -------------------------------------- | ------------------------------ |
| Getting models  | ndb_io.GetModels(ModelName.query(...)) | ModelName.get_multi()          |
| Putting models  | model_pcoll \| ndb_io.PutMulti()       | ModelName.put_multi(models)    |

> | Deleting models | model_pcoll \| ndb_io.DeleteMulti()    | ModelName.delete_multi(models) |

Reviewer: These are confusing because the "Do use" items haven't been introduced yet

Author: Done.

If you think you need to use something else -- e.g. files/etc. -- please talk to @brianrodri/@vojtechjelinek first. Some of these -- typically operations defined by datastore_services -- don't work on Apache Beam in production, while they might work locally.

---

When testing jobs, always inherit from `JobTestBase` and override the class constant `JOB_CLASS`. You can then run `self.assert_job_output_is(...)` or `self.assert_job_output_is_empty()` to verify behavior.

When testing `PTransform`s and `DoFn`s, always inherit from `PipelinedTestBase`.  You can use `self.assert_pcoll_is(...)` or `self.assert_pcoll_is_empty(...)` to verify behavior.

> * **IMPORTANT:** Only one `assert_(job|pcoll)_is` assertion can be performed in a test body. Multiple calls will result in an exception instructing you to split the test apart.

Reviewer: I think it might be worth fleshing out this section some more to describe good testing practices. Maybe provide examples? Also, we should probably specify whether these are written like usual backend tests and, if so, link to the backend tests wiki page

Author: Done.

`Runner`s provide the `run()` method used to visit every node (`PValue`) in the pipeline's DAG by executing the edges (`PTransform`s) to compute their values.  At Oppia, we use `DataflowRunner` to have our `Pipeline`s run on the [Google Cloud Dataflow service](https://cloud.google.com/dataflow).

## Writing Apache Beam Jobs

> For this section, we'll walk through the steps of implementing a job by writing one: `CountExplorationStatesJob`.

Reviewer 3: Explain what your example job does, so that the contributor knows what they're trying to do and what the problem statement actually is. There should probably be some planning stage at the outset for "how" to do it / how to think through the impl, maybe you can guide them through the general thinking that would be needed as step 0.

Author: Done.

Here's the DAG we'll try to implement:

    .--------------. Count states .--------. Sum .-------.
    | Explorations | -----------> | Counts | --> | Total |
    '--------------'              '--------'     '-------'

### 1. Subclass the `base_jobs.JobBase` class and override the `run()` method

Make sure your job class name is clear and concise, because the name is presented to release coordinators:

> <!-- TODO(#11475): Replace this screenshot with one that has job names which follow our naming convention. -->

Reviewer 3: I don't see this tracked in #11475.

It's also on a separate repo so maybe better to just fix it (through in-browser html manipulation) and get rid of the todo?

Author: Done.

Job names should follow the convention: `<Verb><Noun>Job`.

-   For example:

    ```python
    class WeeklyDashboardStatsComputationJob(base_jobs.JobBase):
        """BAD: Name does not begin with a verb."""

        def run(self):
            ...

    class ComputeStatsJob(base_jobs.JobBase):
        """BAD: Unclear what kind of stats are being computed."""

        def run(self):
            ...

    class CountExplorationStatesJob(base_jobs.JobBase):
        """GOOD: Name starts with a verb and "exploration states" is unambiguous."""

        def run(self):
            ...
    ```

Module names should follow the convention: `<noun>_<operation>_jobs.py`.

-   For example:
    * `blog_validation_jobs.py`
    * `dashboard_stats_computation_jobs.py`
    * `exploration_indexing_jobs.py`
    * `exploration_stats_regeneration_jobs.py`
    * `model_validation_jobs.py`

    However, you should always prefer placing jobs in pre-existing modules if an appropriate one already exists.

For this example, we will write our job in the module: `core/jobs/batch_jobs/exploration_inspection_jobs.py`.

### 2. Override the `run()` method to operate on `self.pipeline`

> As illustrated in the Architecture section, jobs are organized by `Pipeline`s, `PTransform`s, and `PCollection`s. Thus, when writing your job, **you must build _off_ of your pipeline** in order to construct the DAG. We can represent this by adding a special `Pipeline` node signifying the _true_ beginning of the DAG:

Reviewer 3: I don't really know what "my pipeline" is. Can you explain conceptually what's happening here and why the pipeline node is the true beginning of the DAG? Does my pipeline already exist or is it a new thing that I am constructing as part of implementing the job? (The current doc suggests the former.)

Author: Done.

      ____________
     /           /\
    |  Pipeline |  |
     \___________\/
            |
            | GetModels()
            v
    .--------------. Count states .--------. Sum .-------.
    | Explorations | -----------> | Counts | --> | Total |
    '--------------'              '--------'     '-------'

Let's see how this would translate into code, starting with the Explorations.

```python
from core.jobs import base_jobs
from core.jobs.io import ndb_io
from core.platform import models

(exp_models,) = models.Registry.import_models([models.NAMES.exploration])

class CountExplorationStatesJob(base_jobs.JobBase):

    def run(self):
        exp_model_pcoll = (
            self.pipeline
            | 'Get all ExplorationModels' >> ndb_io.GetModels(
                exp_models.ExplorationModel.get_all(deleted=False))
        )
```

Observe that:
1. We're using `self.pipeline` as if it were a `PValue`
2. We're using `ndb_io.GetModels` rather than `get_multi`
3. We're passing a Query to `ndb_io.GetModels`

> It just so happens that `Pipeline` does indeed count as a `PValue`, with the special interface "`PBegin`". `PBegin` values signify that the value is never the output of a `PTransform`. Operations can branch off of `PBegin` values, like `ndb_io.GetModels()` does, to begin execution on a pipeline.

Reviewer 3: "It just so happens" suggests coincidence/chance. Instead, try to explain how readers can think of Pipelines as PValues conceptually -- e.g. "Note that a `Pipeline` is a special case of a `PValue` that ..."

Author: Done.

| NDB function           | `PTransform` analogue                        |
| ---------------------- | -------------------------------------------- |
| `models = get_multi()` | `model_pcoll = ndb_io.GetModels(Query(...))` |
| `put_multi(models)`    | `model_pcoll \| ndb_io.PutMulti()`           |
| `delete_multi(keys)`   | `key_pcoll \| ndb_io.DeleteMulti()`          |

Note that `get_multi` has the biggest change in interface, in that it takes a `Query` argument. You can get a query for any model by using the class method `get_all`.
-   **IMPORTANT:** Never use `datastore_services.query_everything()`!! Due to a limitation in Apache Beam, this operation is incredibly slow and inefficient! **You are almost certainly doing something wrong if you need this function.** Ask @brianrodri/@vojtechjelinek for help if you believe you need to use it regardless.

Why should we use these `PTransform` over the simpler `get`/`put`/`delete` functions? **Performance**. The `get`/`put`/`delete` function calls are all *synchronous*, so your job's performance will suffer greatly by waiting for the operations to complete.

> `PTransform`s, on the other hand, are specially crafted to take advantage of the Apache Beam framework and guarantee better performance. In general, **you should always prefer `ndb_io` over any of the `get`/`put`/`delete` functions!** If you think you have a valid need for avoiding `ndb_io`, then speak with @brianrodri/@vojtechjelinek first.

Reviewer 3: Great caveats :)

Author: Ack'd!

Let's get back to implementing the job.

```python
def run(self):
    exp_model_pcoll = (
        self.pipeline
        | 'Get all ExplorationModels' >> ndb_io.GetModels(
            exp_models.ExplorationModel.get_all(deleted=False))
    )

    state_count_pcoll = (
        exp_model_pcoll
        | 'Count states' >> beam.Map(self.get_number_of_states)
    )
```

> Note that we chose to use `beam.Map` here instead of `beam.ParDo`. This is mostly a stylistic choice, as `beam.Map` is just a specialized version of `ParDo`, in that `Map` simply takes each input element and "maps" them to a single output element. In our case, each `ExplorationModel` will map to a single `int`, the number of states.

Reviewer 3: them --> it

Author: Done.

> ```python

Reviewer 3: Introduce this: "Here is the implementation of get_number_of_states. This function just transforms the model into ..." (and you can drop the one-sentence paragraph after the code snippet).

Author: Done.

def get_number_of_states(self, exp_model: ExplorationModel) -> int:
    exp = exp_fetchers.get_exploration_from_model(exp_model)
    return len(exp.states)
```

Our implementation is simple, we transform the model into a domain object and measure the length of its `states` dictionary.

> Finally, we need to sum all the counts together. Fortunately, Apache Beam already has a `PTransform` that does exactly this:

Reviewer 3: I don't immediately see what this PTransform in question is when looking at the code. Maybe move the 'CombineGlobally' part of the explanation up here.

Author: Done.

    state_count_pcoll = (
        exp_model_pcoll
        | 'Count states' >> beam.Map(self.get_number_of_states)
    )

    state_count_sum_pcoll = (
        state_count_pcoll
        | 'Sum values' >> beam.CombineGlobally(sum)
    )
```

`CombineGlobally` uses the input function to combine the values in a `PCollection`. It returns a single-element `PCollection`, one which holds the result of the combination.

With this, our objective is complete. However, there's still more code to write!

### 3. Have the `run()` method return a `PCollection` of `JobRunResult`s

* In English, this means that **the job _must_ report _something_ about what occurred during its execution.** For example, this can be the errors it discovered or the number of successful operations it was able to perform. **Empty results are forbidden!**

  * If you don't think your job has any results worth reporting, then just print a "success" metric with the number of models it processed.

* `JobRunResult` outputs should answer the following questions:

  * Did the job run without any problems? How and why do I know?
  * How much work did the job manage to do?
  * If the job encountered a problem, what caused it?

> Our job is trying to report the total number of states across all explorations, so we need to create `JobRunResult` that holds that information. For this, we can simply use the `as_stdout` helper method:

Reviewer 3: create --> create a

Author: Done.

            exp_models.ExplorationModel.get_all(deleted=False))
    )

    state_count_pcoll = (
        exp_model_pcoll
        | 'Count states' >> beam.Map(self.get_number_of_states)
    )

    state_count_sum_pcoll = (
        state_count_pcoll
        | 'Sum values' >> beam.CombineGlobally(sum)
    )

    return state_count_sum_pcoll | beam.Map(job_run_result.JobRunResult.as_stdout)
```

The method maps every element in a `PCollection` into a `JobRunResult` with the stringified-value as its `stdout`.

With this, our job is fully written! Move on to the next section for instructions on _testing_ your new job.

## Testing Apache Beam Jobs

We'll assume that we've written a fully operational `CountExplorationStatesJob`, and go over the steps in ensuring the job works as intended.

### 1. Inherit from `JobTestBase` and override the class constant `JOB_CLASS`

The current convention is to name your test cases `<JobName>Tests`, but you can create better names if you want to break tests up into groups of specialized focus. For our example, we'll keep things simple.

```python
class CountExplorationStatesJobTests(test_jobs.JobTestBase):

    JOB_CLASS = CountExplorationStatesJob
```

> ### 2. Run assertions using a `assert_job_output_is_*` method

Reviewer 3: Explain best practices for figuring out which tests to write.

Explain what sorts of things can and cannot be tested. How will we verify the parts that cannot be tested through this approach?

Author: Done.

Reviewer 3: Hm. How about the second part? 

> Explain what sorts of things can and cannot be tested. How will we verify the parts that cannot be tested through this approach?

What do these tests guarantee -- will they guarantee a successful and correct prod run? If not, what else is needed?

Author: How's this?

```python
def test_empty_datastore(self):
    # Don't add any explorations to the datastore.
    self.assert_job_output_is_empty()

def test_single_exploration(self):
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e1', 'o1', ['A', 'B', 'C'], ['TextInput'])

    self.assert_job_output_is([
        job_run_result.JobRunResult(stdout='3'),
    ])

def test_many_explorations(self):
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e1', 'o1', ['A', 'B', 'C'], ['TextInput'])
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e1', 'o1', ['D', 'E', 'F', 'G', 'H'], ['TextInput'])
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e1', 'o1', ['I', 'J'], ['TextInput'])
    self.save_new_linear_exp_with_state_names_and_interactions(
        'e1', 'o1', ['K', 'L', 'M', 'N'], ['TextInput'])

    self.assert_job_output_is([
        job_run_result.JobRunResult(stdout='14'),
    ])
```

> Note that `self.assert_job_output_is(...)` and `self.assert_job_output_is_empty()` do as advertised, it runs the job to completion and verifies the result.

Reviewer 3: ... as advertised -- they run the job to completion and verify the result.

Author: Done.

        | 'Sum values' >> beam.CombineGlobally(sum)
    )

    return (
        state_count_sum_pcoll
        | 'Map as stdout' >> beam.Map(job_run_result.JobRunResult.as_stdout)
    )
```

> The method maps every element in a `PCollection` to a `JobRunResult` with their stringified-values as `stdout`.

Reviewer 3: Missing "its" before "stdout", perhaps? (I'm not really sure what the sentence as-written means.)

Author: Done.

Note that `self.assert_job_output_is(...)` and `self.assert_job_output_is_empty()` do as advertised -- they run the job to completion and verify the result.

> **IMPORTANT:** Only one `assert_job_output_is` assertion can be performed in a test body. Multiple calls will result in an exception instructing you to split the test apart.

> Just because a job passes in unit tests does not guarantee it will pass in production. This is because workers, which execute the pipeline code, are run in a special environment where the code base is configured differently. While the Apache Beam jobs team works to cut down on the differences, be careful about using complex and/or confusing objects. The simpler your job, the greater chance it'll work in production!

Reviewer 3: Maybe say who the "Apache Beam jobs team" is,  otherwise this seems like the Apache Beam project's maintainers.

Reviewer 3: Can you also explain or give examples of what sorts of things might fail in this way?

Author: Hmm I don't think it's something worth sharing, they're tiny environmental differences that won't mean anything to job writers so I don't think there's any value in listing examples.

Reviewer 3: Let me put it another way -- what are the steps that are needed to guarantee that things will work in prod? What instructions should the developer give the release team / testing team? Are there certain types of jobs that are known to definitely work?

We can't really have the flow be "write a job and it may or may not work properly" ... we would need some sort of process to introduce new jobs into the codebase, and this doc should be clear on what that process is. Maybe we can discuss tomorrow when meeting with @vojtechjelinek.

Author: As discussed offline, this is not valuable to job writers. Instead, I've added a section instructing job writers to have their jobs tested on production, and linking to the appropriate details page.
